### PR TITLE
[RFC] Move instance meta file under  dir and output novmm log

### PR DESCRIPTION
**Pull request by hamo**
This patchset first moves all one instance related file from .novm/instances/ to .novm/instances/$PID/, including meta file, the writable dir (rename from $PID to writable) and a new log file output from novmm named novmm.log.

Then it enables novmm to output its log to the file .novm/instances/$PID/novmm.log, this will really help debugging.

Now `novm list` function is broken since I changed the dir structure. Will push next commit to fix it.

[Review on Reviewable](https://reviewable.io/reviews/google/novm/25)

### CODE REVIEW

File: novm/manager.py [hunk 1]
```diff
@@ -227,7 +227,7 @@ def state(output):
                 index=1+len(nics)+len(disks),
                 pci=not(nopci),
                 tag="root",
-                tempdir=self._instances.file(str(os.getpid())),
+                tempdir=self._myInstance.file("writable"),
```
Comment (amscanne) [inline novm/manager.py]: It doesn't seem like self._myInstance will be defined when this runs, unless I'm misunderstanding?

File: novm/manager.py [hunk 2]
```diff
@@ -356,8 +360,11 @@ def run_novmm(self,
             # Add extra (debugging) arguments.
             args.extend(["-%s" % x for x in vmmopt])
 
+            # Add log file.
```
Comment (amscanne) [inline novm/manager.py]: I'd like to keep the output if --nofork is specified.

How about something simple like this?

```
if not nofork:
   args.extend["-logfile=%s" % self._instances.file("%d.log" % os.gettpid())])
```

File: src/novmm/main.go [hunk 3]
```diff
@@ -145,6 +146,15 @@ func main() {
 	// Parse all command line options.
 	flag.Parse()
 
+	if *logfile != "" {
```
Comment (amscanne) [inline src/novmm/main.go:149]: Good with this.

**Comment by amscanne** [comment]
I'm generally good with this.

Unless I'm mistaken, I think there's a small problem with the self._myInstance member. I'm fine with either fixing it via my suggestion or if you have some other approach.

File: novm/manager.py (hunk 1 above)
Comment (hamo) [inline novm/manager.py]: self._myInstance is added at line 338~341, since state is a callback function, at this time, we should have _myInstance defined.

**Comment by hamo** [comment]
@amscanne  novm list is fixed with last commit, I have tested with novm list and novm kernels which covers both old data structure and new one.

File: novm/manager.py (hunk 1 above)
Comment (amscanne) [inline novm/manager.py]: Ah I see.

I suppose a nit that I have is the manager class was intended to be able to use programatically for multiple instances.

```
   m = novm.NovmManager()
   m.create(...)
   m.run(...)
```

And I'd like it to be as stateless as possible. With the self._myInstance attribute, that possibility is gone.

Do you think we could just name the log file "%s.log" % instance_id?

That way you get the following files:

   1235.json
   1235/
   1235.log

You would just need to tweak the Nodb.file() method slightly. I think that would be preferable.

(Also, one more favor to ask -- please collapse the commits into working incremental chunks. It helps with bisecting down the road.)

Comment (hamo) [inline novm/manager.py]: Welcome back.

Yes. I will push a new commit to change log file from novmm.log to $PID.log.

And After you review, I will collapse all commits into one and resend a new PR.

**Comment by hamo** [comment]
@amscanne  I have rebased all commits and squash them into one commit. Please kindly review.

File: novm/manager.py (hunk 1 above)
Comment (amscanne) [inline novm/manager.py]: Sorry, I may have been ambiguous.

The self._myInstance stuff is a no-go. The NovmManager() should be useable within other scripts, and may be used to create and manage multiple "containers" simultaneously.

You could make the instance a parameter to the state() closure, and just a regular variable below.

**Comment by hamo** [comment]
I think I finally get your idea. This commit removes self._myInstance and rebased to your last master commit.

**Pull request closed without merging**